File: clients/javascript/__tests__/features/undefined.spec.js
import {default as Invalid} from "../../src/errors/invalid";

require("../support/acceptance");

const Proc = require("../../index");
const client = Proc.connect(process.env.SECRET);

test("does not pass undefined input to calls", async () => {
  await expect(() => client.call("core.echo")).rejects.toThrowError(Invalid);
  await expect(() => client.call("core.echo")).rejects.toThrowError("proc `core.echo' does not accept an undefined input");
});

test("does not pass undefined input to compositions", async () => {
  const composition = client.core.echo.compose(client.core.echo);

  await expect(() => composition.call()).rejects.toThrowError(Invalid);
  await expect(() => composition.call()).rejects.toThrowError("proc `core.echo' does not accept an undefined input");
});

test("does not confuse null with undefined", async () => {
  expect(await client.call("core.echo", null)).toEqual(null);
});

File: clients/ruby/spec/features/undefined_spec.rb
# frozen_string_literal: true

require_relative "support/context/acceptance"

RSpec.describe "ruby client api: undefined" do
  include_context "acceptance"

  it "does not pass undefined input to calls" do
    expect {
      client["core.echo"].call
    }.to raise_error(Proc::Invalid, "proc `core.echo' does not accept an undefined input")
  end

  it "does not pass undefined input to compositions" do
    composition = client.core.echo >> client.core.echo

    expect {
      composition.call
    }.to raise_error(Proc::Invalid, "proc `core.echo' does not accept an undefined input")
  end

  it "does not confuse nil with undefined" do
    expect(client["core.echo"].call(nil)).to be(nil)
  end
end
